### DISCUSSION TITLE
Interoperability with blueprints

### DISCUSSION THREAD

**Discussion by Dejon** [nxrighthere/UnrealCLR#115] [Help] [answered]
I still can't get my head around mapping blueprint objects in the scene to a C# actor object. Are ISystem classes automatically generated, or are they instantiated somewhere that I can't find? Can there only ever one existing at a time? When I look at the BlueprintExtensibility script (as mentioned above, not sure where this gets instantiated) and the corresponding BlueprintActor blueprint, I see where BlueprintActor calls the TestBlueprintActorFunction, passing a reference to itself. While the function is called on a non-static class, and the method is non-static as well, the BlueprintActor would effectively treat it as such as there is no way to determine specific instances of BlueprintExtensibility. I'd really like to give the plugin a whirl, but I need to understand how blueprint instances link to C# code. If anyone could shed some light on this for me, it would be greatly appreciated. Thanks.

**Comment by nxrighthere** [comment, marked as answer]
> Are ISystem classes automatically generated, or are they instantiated somewhere that I can't find? Can there only ever one existing at a time?

It's a custom abstraction in [user code](https://github.com/nxrighthere/UnrealCLR/blob/master/Source/Managed/Tests/Tests.cs#L74).

> When I look at the BlueprintExtensibility script (as mentioned above, not sure where this gets instantiated) and the corresponding BlueprintActor blueprint, I see where BlueprintActor calls the TestBlueprintActorFunction, passing a reference to itself. While the function is called on a non-static class, and the method is non-static as well, the BlueprintActor would effectively treat it as such as there is no way to determine specific instances of BlueprintExtensibility. I'd really like to give the plugin a whirl, but I need to understand how blueprint instances link to C# code.

In this case, an actor and scene component [instantiated](https://github.com/nxrighthere/UnrealCLR/blob/master/Source/Managed/Tests/BlueprintsExtensibility.cs#L22) with `/Game/Tests/BlueprintActor` and `/Game/Tests/BlueprintSceneComponent` blueprints from where [managed functions](https://github.com/nxrighthere/UnrealCLR/blob/master/Source/Managed/Tests/BlueprintsExtensibility.cs#L283) are invoked.

**Comment by Dejon** [comment]
OK, that's much clearer. From what I can see, in all examples the Actor instantiation is performed in managed code. If so, then this doesn't seem very practical, as you'd have to hard-code the values of the initial transform. If I have a scene with blueprint A and blueprint B in it, and each of them wants to have its own BlueprintExtensibility object, what would be the best way to go about that?

**Comment by nxrighthere** [comment]
> If so, then this doesn't seem very practical, as you'd have to hard-code the values of the initial transform.

What do you mean? The object itself in this case is created in the editor, the way how you approach data that related to it is up to you. It might be predefined in the editor or in managed code.

> If I have a scene with blueprint A and blueprint B in it, and each of them wants to have its own BlueprintExtensibility object, what would be the best way to go about that?

`BlueprintsExtensibility` class itself is [instantiated](https://github.com/nxrighthere/UnrealCLR/blob/master/Source/Managed/Tests/Tests.cs#L20) at startup in tests. If you want to instantiate something from blueprints you just have to implement a managed function for that and use [blueprint nodes](https://github.com/nxrighthere/UnrealCLR/blob/master/MANUAL.md#blueprint-functions) of the plugin for invocation.

**Reply by Dejon** [reply]
Yes, I understand that the blueprint is created in the editor, but the code loads the template from file. If I want one or more at different locations in the scene, then the only way that I can see is to instantiate the ISystem class at runtime.

Creating a managed function means creating a class to hold it and, unless the class and method are both static, the blueprint in the scene doesn't have a reference to the instance. Is that clearer? Thanks for your patients, btw.

**Reply by nxrighthere** [reply]
> Yes, I understand that the blueprint is created in the editor, but the code loads the template from file. If I want one or more at different locations in the scene, then the only way that I can see is to instantiate the ISystem class at runtime.

Content of the project can be inspected dynamically with the [`AssetRegistry`](https://github.com/nxrighthere/UnrealCLR/blob/master/API/AssetRegistry.md), so you can load whatever you want based on a path, names, and so on.

> Creating a managed function means creating a class to hold it and, unless the class and method are both static, the blueprint in the scene doesn't have a reference to the instance.

Blueprint nodes don't operate with managed objects and references at all, only with managed functions. If it's possible to pass some references back and forth, then it exists (if valid) in the native engine's memory, the framework of the plugin just represents pointers as references on the managed side.

**Reply by Dejon** [reply]
In the BlueprintExtensibility example, the blueprint can find the managed function because there will only ever be one instance of the class, almost like a singleton. I could add a static method, but I have no way to pass back a reference to the new object to the calling blueprint (i.e. the blueprint has no idea what a BlueprintExtensibility object is). The only way that I can think of implementing this would be to make the BlueprintExtensibility class into a factory so that when the blueprint calls the appropriate static method, it would pass a reference to itself and and the factory would create the required object and forward the blueprint reference on to it. This seems kinda clunky though, and I was wondering if you could give me some advice on how you would approach the problem.

**Reply by nxrighthere** [reply]
> I could add a static method, but I have no way to pass back a reference to the new object to the calling blueprint (i.e. the blueprint has no idea what a BlueprintExtensibility object is).

As I mentioned earlier, blueprint nodes don't operate with managed references (and never will be as long as managed memory is movable).

> This seems kinda clunky though, and I was wondering if you could give me some advice on how you would approach the problem.

Blueprints interoperability is in general not enough flexible at the moment, I'm [working on it](https://github.com/nxrighthere/UnrealCLR/issues/33). My advice is to wait till I finish a new system for advanced use-cases like yours.
